A fix allowing to convert numpy types (np.bool, np.int, np.float) to sympy Float

Trave11er opened the pull request:
The code from #11666 was factored into a separate function (inside sympify.py) that can be called by both sympify and Float. Now can do

Float(np.int64(5))
Float(np.float128(5),12)
Float(np.float32(5.1200000),dps=5)

These previously gave an error but now evaluate to:

5
5.00000000000
5.1200

Corresponding tests have been added in sympy/core/tests/test_numbers.py
Fixes #13275

Trave11er committed:
Float: can now create explicitly from numpy types

The code from #11666 was factored out into a separate function inside
simpify.py and now used by both sympify and Float.
Now issuing
print(Float(np.int64(5)))
print(Float(np.float128(5),12))
print(Float(np.float32(5.1200000),dps=5))
works yielding respectively
5
5.00000000000
5.1200
Also dded unit tests to test_numbers.py by adapting the 'test_numpy'
function from test_simpify.py. However some of the assertions fail.

Trave11er committed:
Added a test for the new code in test_numbers.py

This test was adapted from the one #10666 and now passes successfully
along with the original one in test_sympify.py. However there is an
issue with 'test_issue_10295()'.

Trave11er committed:
Fixed the issue with test_issue_10295()

The convert_np_to_sympy_float was modified so that now all tests
in test_sympify.py and test_numbers.py (incl. new ones) pass for
python 3.5.4.

Trave11er committed:
All tests now also pass for Python 2.7

Trave11er committed:
Removed trailing whitespaces. Test & doctest pass for Python 2.7 & 3.5

Trave11er commented:
Have tried to all the failed tests tests locally and they do succeed. Not sure where to proceed from here - any guidance would be appreciated.

ylemkimon commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,41 @@ def test_Integer_precision():
+
+def test_numpy():
+    from sympy.utilities.pytest import skip
+    import numpy as np
```
You have to do `np = import_module('numpy')`, otherwise it raises `ImportError`.

Trave11er replied:
Thank you - have changed the imports. Regarding Travis - does it meant I have to make sure local dependencies (and their versions) are the same as those used in the automatic tests?

ylemkimon replied:
@Trave11er I meant that it's their problem(s) and not caused by your commits.

ylemkimon reviewed (changes requested):
Other Travis test failures are caused by dependent modules.

Trave11er committed:
Fixed imports to be np = import_module('numpy')

jksuom commented on sympy/core/sympify.py:
```diff
@@ -53,7 +55,7 @@ class CantSympify(object):
 
 def convert_numpy_to_sympy(a):
-    import numpy as np
+    np=import_module('numpy')
```
Add a space on either side of =.

Trave11er replied:
Have done (staged for the next commit)

cbm755 commented on sympy/core/sympify.py:
```diff
@@ -50,6 +52,23 @@ class CantSympify(object):
     pass
 
+# Support for basic numpy datatypes
+
```
two blank lines after previous class/def and no blank line following this comment,

cbm755 replied:
But its not fixed?

Trave11er replied:
Apologies - done now.

cbm755 commented on sympy/core/sympify.py:
```diff
@@ -50,6 +52,23 @@ class CantSympify(object):
+# Support for basic numpy datatypes
+
+def convert_numpy_to_sympy(a):
+    np=import_module('numpy')
```
Someone can call this with a non-numpy object (?).  I think we should assert np is non-None here, as a sanity check

cbm755 commented on sympy/core/tests/test_sympify.py:
```diff
@@ -557,7 +559,7 @@ def equal(x, y):
 
     if not np:
-        skip('numpy not installed.Abort numpy tests.')
+        skip('np not installed.Abort np tests.')
```
?

cbm755 commented on sympy/core/tests/test_sympify.py:
```diff
@@ -2,7 +2,7 @@
     Pow, Or, true, false, Abs, pi, Range)
 from sympy.abc import x, y
-from sympy.core.sympify import sympify, _sympify, SympifyError, kernS
+from sympy.core.sympify import sympify, _sympify, SympifyError, kernS, convert_numpy_to_sympy
```
?

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,41 @@ def test_Integer_precision():
+    if not np:
+        skip('numpy not installed.Abort numpy tests.')
+
+    assert Float(np.bool_(1)) is S(True)
```
I'm not sure we should require `Float(bool)` to work, so no reason for it to work with `np.bool_`.

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,41 @@ def test_Integer_precision():
+        return x == y and type(x) == type(y)
+
+    if not np:
+        skip('numpy not installed.Abort numpy tests.')
```
move up before the `def equal` and fix grammar (space after period).

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -16,7 +16,7 @@
 from mpmath import mpf
 import mpmath
-
+from sympy.external import import_module
```
move down to test function?

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,41 @@ def test_Integer_precision():
+        assert equal(
+            Float(np.int_(12345678912345678)), S(12345678912345678.))
+        assert equal(
+            Float(np.intp(12345678912345678)), S(12345678912345678.))
```
A lot of these tests don't look right to me.  For example here, you can comparing the `S(<int>)` to `S(<float>)`.  They indeed could be the same but its not really what we're supposed to be testing.

Is this whole test a copy-paste with minor edits from sympify?  Maybe that is what we want but all those RHS with `.` are not right.

Trave11er replied:
These tests are indeed a copy-paste from test_sympify. Could you please clarify your comment - my motivation for adding `.` on the RHS was to make sure that the types agree with those on the LHS which all belong to `S(<float>)`.

bmgxyz replied:
I agree with @Trave11er. The LHS is of type ```sympy.core.numbers.Float```, so for the RHS to have the same type, ```S()``` must take ```float``` as an input.

cbm755 replied:
As I tried to explain elsewhere, the RHS should be `Float('12345678912345678', something)`, for some value of `something`.  Bare Python `float`s should be avoided in these tests.

Trave11er replied:
Have rewritten the tests as you outlined in you comment with pseudocode below (to which I also can't reply weirdly). Have changed numbers.py to be the same as committed by @bmgxyz and it passes all the new tests (also, e.g. `Float(np.float64(5),'')` works now).

cbm755 commented on sympy/core/numbers.py:
```diff
@@ -1021,6 +1027,8 @@ def __new__(cls, num, dps=None, prec=None, precision=None):
         if isinstance(num, float):
             _mpf_ = mlib.from_float(num, precision, rnd)
+        elif type(num).__module__ == 'numpy':
+            _mpf_ = mlib.from_float(num, precision, rnd)
```
Eventually we probably want to do this.  But I'm not sure np floats work in all possible paths through the above code for dealing with all the special cases for precision and dps.  For example, does `Float(np...., '')` work?

Trave11er replied:
Will remove if you insist but my suggestion is to keep it allows us to do things like
```
Float(np.float128(5),12)
Float(np.float32(5.1200000),dps=5)
```
Which is better than nothing. The example, you gave indeed fails, but so does `Float(<int>, '')` so it is a more general issue.

cbm755 replied:
I didn't say remove.  I hope the other PR makes those things work.

I assume you mean `Float(<numpy.int>, '')`: please file bug if `Float<sympy int>, '')` isn't working.

cbm755 reviewed (changes requested):
Good ideas here but I'm not sure the Float constructor bits are correct yet.

And test set needs more thought and polish,

Trave11er committed:
Moved import statement and added an assertion, small style fixes

Import statements have been moved into the corresponding functions. An
assertion has been added to make sure that the argument passed to the
numpy-to-sympy converting function is a numpy object.

Trave11er committed:
Removed an assertion, fixed style and reordered relevant tests

Mainly changes in test_numbers.py and test_sympify.py

Trave11er reviewed (commented):
Have incorporated all requested changes, except the code for 1) dealing with special cases of precision and dps 2) '.' on the RHS of assertions in test_numbers.py.

bmgxyz reviewed (commented):
I can't quite speak to the implementation, but the tests seem good to me (especially if they've been copied from ```sympify.py```). I think the final fix for this issue - whatever form it may take - should include most or all of these tests.

smichr commented on sympy/core/sympify.py:
```diff
@@ -50,6 +50,25 @@ class CantSympify(object):
 
+# Support for basic numpy datatypes
+
+def convert_numpy_to_sympy(a):
```
fwiw, I like that this is factored out of the sympify routine.

asmeurer commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1748,43 @@ def test_Integer_precision():
+    assert equal(Float(np.float64(1.1234567891234)),
+                 Float(1.1234567891234, precision=53))
+    assert equal(Float(np.longdouble(1.123456789)),
+                 Float(1.123456789, precision=80))
```
Is longdouble always precision 80? Could it be quad precision?

Trave11er replied:
I believe it is always 80 - the documentation states that longdouble is padded with zeros to either 96 or 128 bits.
Source: https://docs.scipy.org/doc/numpy-dev/user/basics.types.html

cbm755 replied:
I think it is architecture dependent.

Trave11er replied:
@cbm755 you are right - the long double comes from C and is general depends on architecture and compiler. Don't know an easy solution to test for different architectures so removed it along with `np.longcomplex` from test_numbers.py for now.

asmeurer commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1748,43 @@ def test_Integer_precision():
+                 Float(1.123456789, precision=80))
+    assert equal(Float(np.complex64(1 + 2j)), S(1.0 + 2.0 * I))
+    assert equal(Float(np.complex128(1 + 2j)), S(1.0 + 2.0 * I))
+    assert equal(Float(np.longcomplex(1 + 2j)), S(1.0 + 2.0 * I))
```
These should test the precision. Also, I don't know if there are discrepancies in any of these types across platforms, but if there are, we should guard against them.

asmeurer replied:
Actually, it seems kind of odd that Float should accept a complex number. `Float(1j)` currently doesn't work. I think these should actually be raising an exception, at least until https://github.com/sympy/sympy/pull/12192 is finished. You can test that something raises an exception with `raises`.

Trave11er replied:
To test the precision we would need Float instead of S on the RHS but as @asmeurer noticed Float does not currently accept complex numbers. It does accept them through numpy-to-sympy converting function though - so maybe add a raise in Float there whenever np.complex gets passed as an argument?

Trave11er replied:
Added raise for np.complex64 and np.complex128.

cbm755 commented:
For some reason there is no reply box, but this is a reply to...

> Could you please clarify your comment

I mean the stuff on the RHS of your tests would come from Rationals or strings rather than Python `float`.  That is, more like `Float(S(1)/3, precision=...)`.

My suggestion for tests looks something like this in pseudocode:
````
a = numpy.float32(1)/3
b = Float(a)
check that a.precision is same as b.precision
c = Float(S(1)/3, precision=b.precision)
check that relative difference between b and c is less than 2^{-precision}
````
Note, no hardcoded precisions here.  Then repeat that for .float64 and .float128.

For np.float128, also want:
````
a = numpy.float128(1)/3
b = Float(a, '')
check that b.precision is 53.
c = Float(S(1)/3, precision=b.precision)
check that b == c
````

And:
````
a = numpy.float128(1)/3
b = Float(a, precision=10)
check that b.precision is 10.
c = Float(S(1)/3, precision=b.precision)
check that b == c
````

Finally, I'm not convinced we need to test all the non-floating point numpy types here.  At least the non-floating-point-numpy should be in a different `test_` function.

cbm755 commented:
Ideally, Float(numpy.float128) would be loseless: it would have the same binary representation.  I'm not sure if `convert_numpy_to_sympy` does that yet or not.  But that could be a different issue/PR.  It would need its own (careful) test.

Trave11er committed:
Replaced the tests check Float(np.float) implementation

Precision and values of np.float32, np.float64 and np.float128 are
checked. Added 'Raises' for np.complex (not yet implemented).

cbm755 commented on sympy/core/numbers.py:
```diff
@@ -975,6 +977,8 @@ def __new__(cls, num, dps=None, prec=None, precision=None):
         if dps is None and precision is None:
             dps = 15
+            if type(num).__module__ == 'numpy':
+                return convert_numpy_to_sympy(num)
```
Perhaps this is dead code?  Can it be a numpy type when we get here?

Trave11er replied:
Totally agree - removed.

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -18,7 +18,6 @@
 import mpmath
 
 
-
```
unrelated change

Trave11er replied:
Reverted.

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1748,75 @@ def test_Integer_precision():
+            assert(abs(fromnumpy-fromS)/fromS<2**(-fromnumpy._prec+1))
+
+    from sympy.external import import_module
+    np = import_module('numpy')
```
abort when np is None?

Trave11er replied:
Added an assertion.

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1748,75 @@ def test_Integer_precision():
+    check_approximate_equality(a3, testnum)
+
+    fromnumpy = Float(a3)
+    fromS = Float(S(testnum), precision = fromnumpy._prec)
```
`kwarg=a` without space for keyword arguments.  With space for assignment: `fromS = Float...` is correct.

Trave11er replied:
Thank you for explaining - fixed.

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1748,75 @@ def test_Integer_precision():
+
+    raises(TypeError, lambda: Float(np.complex64(1+2j)))
+    raises(TypeError, lambda: Float(np.complex128(1+2j)))
+'''
```
??

Trave11er replied:
`Float(np.complex64(1j))` gives out an error as does `Float(1j)` acting on suggestion from @asmeurer above.

Trave11er committed:
Minor: style fixes and removing dead code

Trave11er committed:
Added 'skip' if not np in test_numbers.py

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
+            fromnumpy = Float(numpyval)
+            fromS = Float(S(testnum), precision = fromnumpy._prec)
+            assert (np.finfo(numpyval).nmant == fromnumpy._prec)
+            assert(abs(fromnumpy-fromS)/fromS<2**(-fromnumpy._prec+1))
```
`/fromS` also goes inside abs.  too much indenting here.  define this after you know you have np.

Trave11er replied:
Yes, makes sense.

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
+
+    fromnumpy = Float(a3)
+    fromS = Float(S(testnum), precision=fromnumpy._prec)
+    assert(fromnumpy._prec == 63)
```
maybe it would be safer just to assert its `> 53`?

Trave11er replied:
Yes, but don't we want to be more strict? As I understand, `np.float128` should certainly have > 52 since `np.float64` has 52 bits in mantissa.

cbm755 replied:
We discussed this already: np.float128 is 63 bits on my computer, but I don't know what the minimum guaranteed is.  If you find a reference we can assert the minimum based on that.

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
+    fromnumpy = Float(a3)
+    fromS = Float(S(testnum), precision=fromnumpy._prec)
+    assert(fromnumpy._prec == 63)
+    assert(fromnumpy == fromS)
```
remove this (too strict I think)

Trave11er replied:
Agreed. Would it be worth adding instead `assert same_and_same_prec(y, Float(S(2)/3, precision=53))`

cbm755 replied:
I don't think that is true.  y should have precision 63 (on my intel cpu).

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
+    fromnumpy = Float(a3, precision=10)
+    fromS = Float(S(testnum), precision=fromnumpy._prec)
+    assert(fromnumpy._prec == 10)
+    assert(fromnumpy == fromS)
```
here you could use `same_and_same_precision`

Trave11er replied:
Totally agree.

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
+
+    raises(TypeError, lambda: Float(np.complex64(1+2j)))
+    raises(TypeError, lambda: Float(np.complex128(1+2j)))
+
```
extra blank line at end of file.

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1749,3 +1749,39 @@ def test_Integer_precision():
+    a2 = np.float64(testnum)
+    check_approximate_equality(a2, testnum)
+    a3 = np.float128(testnum)
+    check_approximate_equality(a3, testnum)
```
I'd prefer a repeating decimal rather than an integer, to be sure the relative error check happens.

maybe:
````
    check_precision_and_relerr(np.float32(2)/3, S(2)/3)
    check_precision_and_relerr(np.float64(2)/3, S(2)/3)
    check_precision_and_relerr(np.float128(2)/3, S(2)/3)
````

Trave11er replied:
Agreed - thank you for clarifying.
